### ISSUE TITLE
Robustness to irrelevant overflows for AnalyticalSimulations

### ISSUE THREAD

**Issue by mjowen** [myokit/myokit#1059]
During the update step for the HH Analytical Simulation (the dynamically defined HHmodel._function), an overflow during the timescale/tau calculation leads to nan outputs in the current and the states for `_t=0`, even though at `_t=0` the states are already known.

Below is an example of a HHmodel._function which can cause this issue (I've simplified it from a larger example where I found this bug but it should still be a viable _function)
``` python
import numpy

def _f(_y0, _t, V, p1, p2, p3):
    _y = [0]*1
    pace = -80.0
    ikur_ui_inf = p1
    ikur_ui_tau = abs(1.0 / (1 + numpy.exp((V + p2) / p3)))
    EK = -70
    _y[0] = ui = ikur_ui_inf + (_y0[0] - ikur_ui_inf) * numpy.exp(-_t / ikur_ui_tau)
    _i = ui * (V - EK)
    return _y, _i

_f(numpy.array([0.5]), numpy.array([0,1]), -80, 0.75, -100, -0.05)
```

If an overflow is triggered in the numpy.exp for the `ikur_ui_tau` statement, then `ikur_ui_tau=0.0`
During the state update, `numpy.exp(-_t / ikur_ui_tau)` returns a nan if `_t==0` and 0 if `t>0` (returns inf for `t<0` which is why the nan is here).
This nan is then returned in the current and states from myokit and this will happen at every `t=0` which means every change in voltage.

Since we already know the state at `_t=0` we could include a line in _f to set _y to _y0, ie.
``` python
    _y[0] = ui = ikur_ui_inf + (_y0[0] - ikur_ui_inf) * numpy.exp(-_t / ikur_ui_tau)
    _y[0][_t==0] = ui[_t==0] = _y0[0]
```

**Comment by MichaelClerx** [comment]
Thanks @mjowen !

> During the state update, numpy.exp(-_t / ikur_ui_tau) returns a nan if _t==0 and 0 if t>0 (returns inf for t<0 which is why the nan is here).

I don't quite understand. If `ikur_ui_tau == 0` it should return `y[0] = [nan, inf, inf, inf...]`, right? So `inf` for `t>0` rather than `0`?

**Comment by mirams** [comment]
That's what it does return, but it could return `y[0] = [initial condition, inf, inf, inf...]` which would let an objective function be calculated as Inf (and used as a big number), rather than NaN (not usable!).

**Comment by mjowen** [comment]
There's a minus sign in front of the _t. So for `_t=1`, you have `numpy.exp(-1/0.0)=numpy.exp(-inf)=0`, ie immediately jump to steady state

**Comment by MichaelClerx** [comment]
Fair enough! 
I'm open to PRs that implement this, particularly if they contain unit tests to make sure we don't lose this feature if/when the HH analytical simulations get rewritten as pure C.
If it's up to me it'll be after May 7th I'm afraid 😅

**Comment by MichaelClerx** [comment]
(And I know that in this context you definitely want to run with these parameters, but in any other context I would advise you to place better boundaries/priors on p3 :D )

**Comment by MichaelClerx** [comment]
Closed in #1060
